### PR TITLE
JAV-438 update execution result query controller for graph

### PR DESCRIPTION
**Pull request by zhengyangyong**
Signed-off-by: zhengyangyong <yangyong.zheng@huawei.com>
修改了结果查询接口，更容易绘图

### CODE REVIEW

File: saga-spring/src/main/java/io/servicecomb/saga/spring/SagaController.java [hunk 1]
```diff
@@ -125,36 +148,90 @@ public SagaController(SagaExecutionComponent sagaExecutionComponent, SagaEventRe
...
+          point.add(child.value().id());
+          loopLoadGraphNodes(router, child);
+        }
+      }
+    }
+  }
+
+  private boolean isNodeValid(Node<SagaRequest> node) {
+    return !NoOpSagaRequest.SAGA_END_REQUEST.id().equals(node.value().id())
+        && node.children() != null && node.children().size() != 0;
```
Comment (seanyinx) [inline saga-spring/src/main/java/io/servicecomb/saga/spring/SagaController.java]: only saga end request can have empty children. empty and null checking are duplicate with end request id checking

File: saga-spring/src/main/java/io/servicecomb/saga/spring/SagaController.java [hunk 2]
```diff
@@ -182,17 +263,41 @@ public SagaRequestQueryResult(int pageIndex, int pageSize, int totalPages, List<
...
-  protected static class SagaRequest {
+  static class SagaExecution {
     public long id;
+    public String sagaId;
     public long startTime;
     public String status;
     public long completedTime;
 
-    public SagaRequest(long id, long startTime, long completedTime, String status) {
+    public SagaExecution() {
```
Comment (seanyinx) [inline saga-spring/src/main/java/io/servicecomb/saga/spring/SagaController.java]: why no arg constructor if the class is only for deserialization?

File: saga-spring/src/main/java/io/servicecomb/saga/spring/SagaController.java [hunk 3]
```diff
@@ -182,17 +263,41 @@ public SagaRequestQueryResult(int pageIndex, int pageSize, int totalPages, List<
...
     }
   }
+
+  @JsonAutoDetect(fieldVisibility = ANY, getterVisibility = NONE, setterVisibility = NONE)
+  static class SagaExecutionDetail {
+    public Map<String, List<String>> router;
+    public Map<String, String> status;
+    public Map<String, String> error;
+
+    public SagaExecutionDetail() {
```
Comment (seanyinx) [inline saga-spring/src/main/java/io/servicecomb/saga/spring/SagaController.java]: why no arg constructor if the class is only for deserialization?

File: saga-spring/src/main/java/io/servicecomb/saga/spring/SagaController.java [hunk 4]
```diff
@@ -166,14 +243,18 @@ private SagaEventVo(long id, String sagaId, long creationTime, String type, Stri
...
 
   @JsonAutoDetect(fieldVisibility = ANY, getterVisibility = NONE, setterVisibility = NONE)
-  protected static class SagaRequestQueryResult {
+  static class SagaExecutionQueryResult {
     public int pageIndex;
     public int pageSize;
     public int totalPages;
 
-    public List<SagaRequest> requests = null;
+    public List<SagaExecution> requests = null;
```
Comment (seanyinx) [inline saga-spring/src/main/java/io/servicecomb/saga/spring/SagaController.java]: fields can all be final without no arg constructor

File: saga-spring/src/main/java/io/servicecomb/saga/spring/SagaController.java [hunk 5]
```diff
@@ -182,17 +263,41 @@ public SagaRequestQueryResult(int pageIndex, int pageSize, int totalPages, List<
...
+      this.sagaId = sagaId;
       this.startTime = startTime;
       this.completedTime = completedTime;
       this.status = status;
     }
   }
+
+  @JsonAutoDetect(fieldVisibility = ANY, getterVisibility = NONE, setterVisibility = NONE)
+  static class SagaExecutionDetail {
+    public Map<String, List<String>> router;
```
Comment (seanyinx) [inline saga-spring/src/main/java/io/servicecomb/saga/spring/SagaController.java]: fields can all be final without no arg constructor

File: saga-spring/src/main/java/io/servicecomb/saga/spring/SagaController.java [hunk 6]
```diff
@@ -125,36 +148,90 @@ public SagaController(SagaExecutionComponent sagaExecutionComponent, SagaEventRe
...
-        .forEach(e -> events.add(new SagaEventVo(e.id(), e.sagaId(), e.creationTime(), e.type(), e.contentJson())));
-    return ResponseEntity.ok(events);
+  public ResponseEntity<SagaExecutionDetail> queryExecutionDetail(@PathVariable String sagaId) {
+    SagaEventEntity[] entities = repo.findBySagaId(sagaId).toArray(new SagaEventEntity[0]);
+    Optional<SagaEventEntity> sagaStartEvent = Arrays.stream(entities)
+        .filter(entity -> SagaStartedEvent.class.getSimpleName().equals(entity.type())).findFirst();
+    Map<String, List<String>> router = new HashMap<>();
+    Map<String, String> status = new HashMap<>();
+    Map<String, String> error = new HashMap<>();
+    if (sagaStartEvent.isPresent()) {
```
[the remark this answers is not in the document]
Comment (zhengyangyong) [reply saga-spring/src/main/java/io/servicecomb/saga/spring/SagaController.java]: Done

File: saga-spring/src/main/java/io/servicecomb/saga/spring/SagaController.java [hunk 7]
```diff
@@ -125,36 +148,90 @@ public SagaController(SagaExecutionComponent sagaExecutionComponent, SagaEventRe
...
+          .filter(entity -> TransactionAbortedEvent.class.getSimpleName().equals(entity.type())).collect(
+              Collectors.toList());
+      for (SagaEventEntity transactionAbortEvent : transactionAbortEvents) {
+        try {
+          JsonNode root = mapper.readTree(transactionAbortEvent.contentJson());
+          String id = root.at("/request/id").asText();
+          status.put(id, "Failed");
+          error.put(id, root.at("/response/body").asText());
+        } catch (IOException ignored) {
+        }
```
[the remark this answers is not in the document]
Comment (zhengyangyong) [reply saga-spring/src/main/java/io/servicecomb/saga/spring/SagaController.java]: Done

File: saga-spring/src/main/java/io/servicecomb/saga/spring/SagaController.java (hunk 1 above)
Comment (zhengyangyong) [reply saga-spring/src/main/java/io/servicecomb/saga/spring/SagaController.java]: Done

File: saga-spring/src/main/java/io/servicecomb/saga/spring/SagaController.java (hunk 2 above)
Comment (zhengyangyong) [reply saga-spring/src/main/java/io/servicecomb/saga/spring/SagaController.java]: 在SagaSpringApplicationTest中做测试有反序列化

File: saga-spring/src/main/java/io/servicecomb/saga/spring/SagaController.java (hunk 5 above)
Comment (zhengyangyong) [reply saga-spring/src/main/java/io/servicecomb/saga/spring/SagaController.java]: 在SagaSpringApplicationTest中做测试有反序列化

File: saga-spring/src/main/java/io/servicecomb/saga/spring/SagaController.java (hunk 3 above)
Comment (zhengyangyong) [reply saga-spring/src/main/java/io/servicecomb/saga/spring/SagaController.java]: 在SagaSpringApplicationTest中做测试有反序列化

File: saga-spring/src/main/java/io/servicecomb/saga/spring/SagaController.java (hunk 4 above)
Comment (zhengyangyong) [reply saga-spring/src/main/java/io/servicecomb/saga/spring/SagaController.java]: 在SagaSpringApplicationTest中做测试有反序列化

File: saga-spring/src/main/java/io/servicecomb/saga/spring/SagaController.java [hunk 8]
```diff
@@ -108,53 +104,34 @@ public SagaController(SagaExecutionComponent sagaExecutionComponent, SagaEventRe
...
+          && Long.parseLong(startTime) <= Long.parseLong(endTime);
+    } catch (NumberFormatException nfe) {
+      return false;
+    }
   }
 
   @ApiResponses({
       @ApiResponse(code = 200, response = String.class, message = "success"),
-      @ApiResponse(code = 400, response = String.class, message = "illegal request content"),
+      @ApiResponse(code = 200, response = String.class, message = "success"),
```
Comment (seanyinx) [inline saga-spring/src/main/java/io/servicecomb/saga/spring/SagaController.java]: seems to be duplicate response

File: saga-spring/src/main/java/io/servicecomb/saga/spring/SagaExecutionQueryService.java [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+      List<String> point = router.computeIfAbsent(node.value().id(), key -> new ArrayList<>());
+      for (Node<SagaRequest> child : node.children()) {
+        point.add(child.value().id());
+        loopLoadGraphNodes(router, child);
+      }
+    }
+  }
+
+  private boolean isNodeValid(Node<SagaRequest> node) {
+    return node.children() != null && node.children().size() != 0;
```
Comment (seanyinx) [inline saga-spring/src/main/java/io/servicecomb/saga/spring/SagaExecutionQueryService.java]: children can never be null, checking `node.children().isEmpty` is enough

**Pull request merged**